internal/engine/planner/struct_matcher.go

```diff
@@ -0,0 +1,447 @@
+				return false, nil
+			}
+			return containsOnlyKnownValues(iterRange), nil
+		},
+	}
+	return processors([]expressionProcessor{s1, s2, s3})
+}
+
+func mkLogicalOr(args []celast.Expr) celast.Expr {
+	const logicalOrArity = 2
+	if len(args) == 1 {
+		return args[0]
+	}
+	if len(args) == logicalOrArity {
+		return internal.MkCallExpr(operators.LogicalOr, args...)
+	}
+	return internal.MkCallExpr(operators.LogicalOr, args[0], mkLogicalOr(args[1:]))
+}
+
+func mkOption(op string, key, val, expr, constExpr celast.Expr) celast.Expr {
+	if op == "" {
+		panic("mkOption: operation is empty")
+	}
+	lhs := internal.MkCallExpr(operators.Equals, expr, key)
+	rhs := internal.MkCallExpr(op, constExpr, val)
+	return internal.MkCallExpr(operators.LogicalAnd, lhs, rhs)
+}
+
+type lambdaMatcher struct {
+	iterRange        celast.Expr
+	innerExpr        celast.Expr
+	rootMatcher      *exprMatcher
+	partialEvaluator *partialEvaluator
+	iterVar          string
+	iterVar2         string
+}
+
+func containsOnlyKnownValues(expr celast.Expr) bool {
+	switch expr.Kind() {
+	case celast.LiteralKind:
+		return true
+	case celast.ListKind:
+		list := expr.AsList()
+		for _, element := range list.Elements() {
+			if !containsOnlyKnownValues(element) {
+				return false
+			}
+		}
+		return true
+	case celast.MapKind:
+		mapExpr := expr.AsMap()
+		for _, entry := range mapExpr.Entries() {
+			me := entry.AsMapEntry()
+			if !containsOnlyKnownValues(me.Key()) || !containsOnlyKnownValues(me.Value()) {
+				return false
+			}
+		}
+		return true
+	case celast.StructKind:
+		st := expr.AsStruct()
+		for _, field := range st.Fields() {
+			if !containsOnlyKnownValues(field.AsStructField().Value()) {
+				return false
+			}
+		}
+		return true
+	default:
+		// For other types like IdentKind, SelectKind, CallKind, ComprehensionKind, etc.
+		// These typically involve variables or complex expressions
+		return false
+	}
+}
+
+func (l *lambdaMatcher) Process(ctx context.Context, e celast.Expr) (bool, celast.Expr, error) {
+	r, err := l.rootMatcher.run(e)
+	if err != nil || !r {
+		return false, nil, err
+	}
+	ep, err := celast.ExprToProto(e)
+	if err != nil {
+		return false, nil, fmt.Errorf("fail to convert expr to proto: %w", err)
+	}
+	ce := ep.GetComprehensionExpr()
+	if ce == nil {
+		return false, nil, nil
+	}
+
+	lambda, err := buildLambdaAST(ce)
+	if err != nil {
+		return false, nil, err
+	}
+
+	if lambda.operator != Exists {
+		return false, nil, err
+	}
+	l.iterRange, err = celast.ProtoToExpr(lambda.iterRange)
+	if err != nil {
+		return false, nil, err
+	}
+	l.innerExpr, err = celast.ProtoToExpr(lambda.expr)
+	if err != nil {
+		return false, nil, err
+	}
+	l.iterVar = lambda.iterVar
+	l.iterVar2 = lambda.iterVar2
+
+	knownVars := make(map[string]any, len(l.partialEvaluator.knownVars)+nLambdaVars)
+	maps.Copy(knownVars, l.partialEvaluator.knownVars)
+
+	const maxItems = 25
```

Comment:
Sorry, I know we talked about 25 being a "reasonable" number. But, now that I think about it, most query languages have an efficient operator for comparing against a list of items so it's probably better to make the cut-off a bit smaller to about 10 items. If the lambda expression is particularly complicated, a 25 of those OR'ed together is probably less efficient to evaluate compared to the equivalent array expression in the target query language.